daemons/attrd/attrd_cib.c

```diff
@@ -485,12 +566,6 @@ attrd_write_attribute(attribute_t *a, bool ignore_delay)
         v->requested = NULL;
         if (v->current) {
             v->requested = strdup(v->current);
-        } else {
-            /* Older attrd versions don't know about the cib_mixed_update
```

Comment:
Looking at this anew, I'm wondering if we had an oversight re: transactions or if I'm forgetting something right now.

If we send a commit-transaction request in a mixed-version cluster, the older nodes may not recognize it... right?

It seems like that would apply to any new CIB request type that we've ever added though, and I don't remember seeing a lot of special handling. In particular, `cib_scope_local` does nothing unless we're in legacy mode.

---

Related: There are still some comments in the transaction code about syncing the result CIB. I don't believe we do that anymore, now that the whole transaction is built on the client side and bundled into one commit-transaction request. The commit-transaction request would just get sent to all nodes to process locally, and to activate on success.

Reply:
Oh wow, I don't think we've had any new CIB request types since 1.1.8 :/

Reply:
Bah, don't know how I overlooked that. We also have to consider a newer CIB client (command-line tool) on a remote node talking to an older cluster node, though maybe it's sufficient to just get a "not supported" error in that case.

My memory is hazy since it changed so much -- I thought the local CIB manager applied transactions to a copy of the CIB, then the result was sync'd to peers (including itself) via existing mechanisms.

Reply:
Currently attrd and execd implement their own protocol versions, and each daemon has its own IPC connection sequence. Eventually I'd like to get to where we have a common connection sequence (handshake) used with all daemons that would include passing key information like feature set and daemon protocol version supported. Not sure what based has currently.

Reply:
I think "not supported" would be fine for a remote client.

It did, in the original implementation where the whole transaction was built in the CIB manager. In the PR where we changed it to build the transaction in the client and send it as one big request with multiple nested requests, we also made it behave like any other "modifies" request. On v2 it gets sent to all nodes to apply locally (assuming I'm remembering the mechanism right) and on v1 it gets sent to the primary or broadcast or whatever.

Reply:
I don't think based has anything like a handshake that gets the feature set. It could be pulled from the CIB XML if that doesn't miss cases. My idea was to mimic the legacy mode determination. If we receive a request with feature set too low to support transactions, then we set a local_transactions boolean to true for as long as the based process remains active. If that's true, then any future transactions get processed locally and synced instead of sending it to all nodes to process. (Or do that based on the feature set in CIB XML.)

Reply:
If it's already a generic modify request by the time it's broadcast, then we don't need to worry about compatibility (other than remotes)

Reply:
It's never a modify request. It's a commit-transaction request, which has the modifies flag and contains an arbitrary sequence of modifying requests within it.

A backward compatible implementation would apply all changes locally and then sync it.

Reply:
In any case it should be able to operate that way -- the local daemon applies the transaction to a copy of the CIB, then generates a diff based on that and sends the diff around as if it were a single modify op

Reply:
> Commit is a local op, and IIRC the local daemon applies the entire transaction, and only the result is sent around. But I could be misremembering.

That was true originally. It changed when I realized we could put the whole transaction into an XML blob that doesn't need to maintain internal state in the local CIB manager.

Reply:
> In any case it should be able to operate that way -- the local daemon applies the transaction to a copy of the CIB, then generates a diff based on that and sends the diff around as if it were a single modify op

That could work -- a new `PCMK__CIB_REQUEST_APPLY_PATCH` request using a patchset generated from the local transaction result CIB. I don't know off hand how much easier or harder that would be than syncing the result CIB. It would probably have to copy or functionize what `sync_our_cib()` does to build a new request.

The only definite advantage would be less messaging overhead for large CIBs -- but that might be substantial.

---

Not sure if that should be THE way to implement transactions, or only compatibility code for when other nodes won't recognize "commit-transaction."

The reason we do it the current way (send the whole transaction request to all nodes to try to commit) is that that's how we treat every other request that modifies the CIB. It's not unreasonable to process a transaction locally and then apply the diff. I had just wanted to avoid special handling unless necessary.

Reply:
A replace instead of a diff would be fine. The goal would be to avoid adding any new ops that need to be sent around. If we can have a single way rather than new/legacy that would be better. Basically commit would be a local op, but it would generate a modify/replace/sync/whatever op that would be normal.

Reply:
Thanks. I'll look into both approaches and see which one seems cleaner.

By the way, the attrd changes in this PR are unlikely to change based on the changes to the transaction implementation (even though we can't merge this until the transaction implementation is fixed).

Reply:
Fixed via #3235